module/state_synchronization/storage.go

```diff
@@ -0,0 +1,957 @@
+	latestIncorporatedHeightValue := make([]byte, 8)
+	binary.LittleEndian.PutUint64(latestIncorporatedHeightValue, startHeight)
+
+	storedDataLowerBoundKey := makeGlobalStateKey(globalStateStoredDataLowerBound)
+	storedDataLowerBoundValue := make([]byte, 8)
+	binary.LittleEndian.PutUint64(storedDataLowerBoundValue, startHeight)
+
+	return s.db.Update(func(txn *badger.Txn) error {
+		if err := txn.Set(latestIncorporatedHeightKey, latestIncorporatedHeightValue); err != nil {
+			return fmt.Errorf("failed to set latest incorporated height: %w", err)
+		}
+
+		if err := txn.Set(storedDataLowerBoundKey, storedDataLowerBoundValue); err != nil {
+			return fmt.Errorf("failed to set stored data lower bound: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (s *Storage) GetLatestIncorporatedHeight() (uint64, error) {
+	latestIncorporatedHeightKey := makeGlobalStateKey(globalStateLatestIncorporatedHeight)
+	var latestIncorporatedHeight uint64
+
+	if err := s.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(latestIncorporatedHeightKey)
+		if err != nil {
+			return fmt.Errorf("failed to find latest incorporated height entry: %w", err)
+		}
+
+		latestIncorporatedHeightValue, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve latest incorporated height value: %w", err)
+		}
+
+		latestIncorporatedHeight = binary.LittleEndian.Uint64(latestIncorporatedHeightValue)
+
+		return nil
+	}); err != nil {
+		return 0, err
+	}
+
+	return latestIncorporatedHeight, nil
+}
+
+// GetStoredDataLowerBound returns the maximum h such that the storage contains no data for any block heights <= h
+// TODO: please help me think of a better name for this function
+func (s *Storage) GetStoredDataLowerBound() (uint64, error) {
+	storedDataLowerBoundKey := makeGlobalStateKey(globalStateStoredDataLowerBound)
+	var storedDataLowerBound uint64
+
+	if err := s.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(storedDataLowerBoundKey)
+		if err != nil {
+			return fmt.Errorf("failed to find stored data lower bound entry: %w", err)
+		}
+
+		storedDataLowerBoundValue, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve stored data lower bound value: %w", err)
+		}
+
+		storedDataLowerBound = binary.LittleEndian.Uint64(storedDataLowerBoundValue)
+
+		return nil
+	}); err != nil {
+		return 0, err
+	}
+
+	return storedDataLowerBound, nil
+}
+
+// Prune removes all data from storage corresponding to block heights lower than or equal to the given height.
+// This includes all blob tree records, blob records, and blob data.
+// This should only be called with height <= latest incorporated height.
+func (s *Storage) Prune(height uint64) ([]cid.Cid, error) {
+	storedDataLowerBoundKey := makeGlobalStateKey(globalStateStoredDataLowerBound)
+	storedDataLowerBoundValue := make([]byte, 8)
+	binary.LittleEndian.PutUint64(storedDataLowerBoundValue, height)
+
+	if err := s.db.Update(func(txn *badger.Txn) error {
```

Comment:
We are updating the database before actually deleting pruning (deleting Cids). If the node crashes right after updating the database, how can we recover from the inconsistency between database and data store?

Reply:
Yes, this is supposed to be handled by `replay`. Can you check that this is handled properly?